### PR TITLE
[DNM] agent: Properly stop the gRPC server

### LINKED ISSUE #445: vsock semantics
# Background

Whilst working on the agent tracing (#415), I noticed that occasionally the agent's gRPC server was not shutting down. This is a big problem for tracing since we need to guarantee a clean agent shutdown. Since the gRPC server is not shutting down, it is stopping the agent from finalising trace spans, meaning only partial trace information is being sent back to the Jaeger agent on the host.

# Findings

From what I can see, the hang has nothing to do with the gRPC library, nor the existing agent code, nor the new agent tracing code: the problem appears to be vsock: toggling `use_vsock = false` in `configuration.toml` stops the hang.

## Observed vsock behaviour in kata-agent

Tracing the agent code shows that deep in the bowels of the gRPC server, the following is happening:

```go
lis, err := vsock.Listen(vSockPort)
err = grpcServer.Serve(lis)
```
And `grpcServer.Serve()` is doing this:

```go
for {
    rawConn, err := lis.Accept()

    s.handleRawConn(rawConn)
}
```

Using a serial device or a unix domain socket, that `Accept()` call will fail when the client (the `kata-runtime`) disconnects, returning an error (a Yamux `ErrSessionShutdown`).

But with `kata-runtime`, vsock seemingly never fails at this point, always returning a new connection. As a result, the infinite loop never exits, ergo the hang.

It may be that we are mis-using vsock somehow or it may be that we're hitting a kernel bug with the vsock module.

# Recreating the problem

To help debugging this issue, I've created a branch that:

- Allows the agent to be run stand-alone (and it also includes a ton of debug messages :)
- Provides a `kata-agent-ctl` command that provides just enough functionality to trigger the problem.

# Agent

- Clone https://github.com/jodh-intel/agent-1/tree/run-on-host to `$GOPATH/src/github.com/kata-containers/agent`.
- Then:
  ```sh
  $ make clean && make
  ```

# Testing

## Test using a unix domain socket

- Start kata-agent (server)

  ```sh
  $ cd $GOPATH/src/github.com/kata-containers/agent
  $ ./kata-agent --debug --grpc-trace --channelPath unix:///tmp/kata-agent.sock --no-udev
  ```
- Start client

  ```sh
  $ cli/kata-agent-ctl --keep-alive --debug --agentAddress unix:///tmp/kata-agent.sock --enable-yamux
  ```

Running the client should display some output and exit. It should also cause the server to exit.

## Test using a vsock socket

- Setup vsock on host

  ```sh
  $ sudo modprobe vhost_vsock
  $ sudo chmod 666 /dev/vhost-vsock /dev/vsock
  ```
- Start qemu with `-device vhost-vsock-pci,id=vhost-vsock-pci0,guest-cid=3`

- Start kata-agent server (inside qemu)

  ```sh
  $ sudo ./kata-agent --debug --grpc-trace --no-udev
  ```

- Start client (on host)

  ```sh
  $ sudo cli/kata-agent-ctl --keep-alive --debug
  ```

  (Note: the `--keep-alive` option doesn't appear to make life better fwics so that *may* be optional).

What you will probably find is that the behaviour is now erratic:

- Sometimes the server will exit when the client exits (as expected).
- Sometimes the server will not exit when the client exits, but re-running the client will cause the server and client to exit.
- Sometimes the server will not exit when the client exits and re-running the client will cause the server to immediately exit, but the client will then hang waiting for the server.

@stefanha - could you tal and sanity check the above? Is there anything obvious we're doing or not doing that might trigger behaviour like this? The agent file dealing with vsock is:

- https://github.com/kata-containers/agent/blob/master/channel.go

/cc @sboeuf, @mcastelino,

### PR DESCRIPTION
**Pull request by sboeuf**
This commit attempts to close cleanly the gRPC server so that tracing
will be ended properly.

Fixes #445

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

### CODE REVIEW

File: agent.go [hunk 1]
```diff
@@ -524,6 +525,36 @@ func (s *sandbox) teardownSharedPidNs() error {
...
+	go func() {
+		s.gracefulStopGRPC()
+		close(done)
+	}()
+
+	select {
+	case <-done:
+		return
+	case <-time.After(timeout):
+		fieldLogger.Warnf("Could not gracefully stop the server after %v", timeout)
```
Comment (devimc) [inline agent.go]: `fieldLogger.WithField("timeout", timeout)....`

File: agent.go [hunk 2]
```diff
@@ -524,6 +525,36 @@ func (s *sandbox) teardownSharedPidNs() error {
...
+	<-s.stopServer
+
+	defer fieldLogger.Info("gRPC server stopped")
+
+	// Try to gracefully stop the server for a minute
+	timeout := time.Minute
+	done := make(chan struct{})
+	go func() {
+		s.gracefulStopGRPC()
+		close(done)
```
[the remark this answers is not in the document]
Comment (sboeuf) [reply agent.go]: Yes let me do that a little bit better

[the remark this answers is not in the document]
Comment (sboeuf) [reply agent.go]: done

[the remark this answers is not in the document]
Comment (egernst) [reply agent.go]: Oh that’s what that smell was....

**Comment by sboeuf** [comment]
CI failures about license check should be fixed by https://github.com/kata-containers/tests/pull/1101

**Comment by sboeuf** [comment]
@jodh-intel 

Ok so I did a bit of testing and unfortunately, this PR breaks Kata Containers... The problem being that by closing the gRPC server, we end up closing the connection with the proxy, since the serial connection is managed through Yamux who is managed through gRPC... The server is properly stopped after all gRPC connections are returning, but this means we have a race where the proxy connection is being closed while returning the answer to the runtime.
So I think we should keep the current behavior of not trying to stop the server, and instead let the agent run.

That being said, the initial purpose of this PR was to be able to stop the span of tracing, and I think it can be done by having the same mechanism of Go channel to stop the span from the main agent go routine, instead of from the `DestroySandbox()` one.
You need to make sure the `DestroySandbox()` will wait for the end of the span before it returns, otherwise the VM might be killed while the whole span is sent out of the VM through vsock.
This means we need to be careful with the use of `defer` statements, as the first defer should be the wait for the end of the span.

**Comment by jodh-intel** [comment]
Hi @sboeuf - thanks for looking into this.

> You need to make sure the DestroySandbox() will wait for the end of the span before it returns, otherwise the VM might be killed while the whole span is sent out of the VM through vsock.

Alas, that won't work. It could wait for the end of any spans *it* creates, but what about *its own span?* One of the main reasons for adding tracing to the agent is to be able to trace the gRPC calls. Hence, `DestroySandbox()`, being a gRPC API actually needs to *fully return* before *its* span can be completed. See the problem? :)

The tracing support assumes vsock, which implies no proxy. So we *could* potentially do the full agent shutdown in the non-proxy case. It isn't ideal having the two code paths. However, the proxy has to be considered "legacy" at some point and there were always going to be conditions on allowing agent tracing.

Shutting down the agent cleanly is the right thing to do imho: it makes perfect sense, whereas the current design is much less clear.

**Comment by jodh-intel** [comment]
Testing with my tracing code shows there are a few problems here:

- The runtime kills the agent too early (before the agent has shutdown and finalised the trace spans).

  That can be resolved by allowing the VM to shut itself down and removing the QMP quit from the runtime.

- Once that's done, the behaviour is racy:

  - if the workload runs "quickly" (`busybox true`), the trace span generally completes (since the agent shuts down correctly after a graceful gRPC server stop).
  - if the workload takes a little longer, the agent hits the timeout, then randomly either the forced-stop works, or it gets stuck waiting on the grpc internal `WaitGroup` issue I've seen.

**Comment by sboeuf** [comment]
@jodh-intel 
>The runtime kills the agent too early (before the agent has shutdown and finalised the trace spans).

True, and that's why we should make sure the agent gets the chance to spit the logs before it ends.

>That can be resolved by allowing the VM to shut itself down and removing the QMP quit from the runtime.

I don't like this as it would make the agent actively responsible for the shutdown of the VM, which should be done by the runtime IMO. And as you mentioned, this leads into race conditions.

I think that we should simply consider `DestroySandbox()` as the last call that can be received by the agent. With this in mind, we would consider doing a few things before we can actually return from this request.
Concretely, this means we have to send a signal from the `DestroySandbox()` implementation to the agent main routine to handle tracing span finalization. When receiving this signal, the agent would close all the spans and send them out. Afterwards, it would simply signal the `DestroySandbox()` thread back, using another channel, and it would return.

Now, if you don't like the fact that we might miss a few logs from `DestroySandbox()` (because we have to do a span.Finish before to send the first signal), then maybe we need to introduce an extra gRPC call called `StopAgent()` or `StopTracing()` that would be in charge of doing this. And this would assure us that the agent actually had the time to send all the traces before it returns.
Actually the more I think about this, the more `StopTracing()` seems to be appropriate since it could be called from the runtime only if we have `vsock`.

WDYT about this proposal?

**Comment by jodh-intel** [comment]
@sboeuf - yes, funnily enough I used to have just such a `StopAgent()` call in a local branch for this reason. I can add it back but it actually just "moves the problem"... because `StopAgent()` is yet another gRPC call which has an associated trace span.

I'll take another look at this design Monday with a clearer head and more coffee on hand...! ;)

**Comment by sboeuf** [comment]
@jodh-intel 
>I can add it back but it actually just "moves the problem"... because StopAgent() is yet another gRPC call which has an associated trace span.

No but you don't want a span in this call, that's the point. Or if you want one, just start it to notify that we reached this function, but then stop it right before we signal the main agent thread.
We have to be okay with loosing a little bit of tracing for the sake of having proper closure here.

One more thing, if we don't want to over complicate this, let's put the fact that we don't stop the agent out the picture here.

**Comment by sboeuf** [comment]
Just to mention this is a test/enhancement to handle proper agent termination, but this is not ready yet and being investigated by @jodh-intel and myself.

**Comment by jodh-intel** [comment]
Incorporated into #415.

**Comment by grahamwhaley** [comment]
@sboeuf @jodh-intel - as this is now in #415 - should we close this PR?

**Comment by jodh-intel** [comment]
Thanks @sboeuf. wfm - closing...

**Pull request closed without merging**